### PR TITLE
[BugFix] Fix count distinct rewrite bug

### PR DESCRIPTION
**Pull request by Seaven**
Fixes https://github.com/StarRocks/starrocks/pull/17698

COUNT rewrite erorr on second stage
error: COUNT -> COUNT + COUNT
right: COUNT -> COUNT + SUM

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 4 / 4 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java [hunk 1]
```diff
@@ -206,7 +209,8 @@ private CallOperator transformOtherAgg(CallOperator origin, List<ScalarOperator>
         } else {
             String secondFuncName = OTHER_FUNCTION_TRANS.get(originFuncName).second;
             Type[] argTypes = args.stream().map(ScalarOperator::getType).toArray(Type[]::new);
-            Function newFunc = origin.getFunction().updateArgType(argTypes);
+            Function newFunc = Expr.getBuiltinFunction(secondFuncName, argTypes,
+                    Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF);
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java:213]: add a Precondition check for newFunc not null?

**Review by packy92 — DISMISSED** [review]

File: test/sql/test_agg_function/T/test_count_distinct [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+insert into test_cc values('a','a', '2022-04-18 01:01:00', 1);
+insert into test_cc values('a','b', '2022-04-18 02:01:00', 2);
+insert into test_cc values('a','a', '2022-04-18 02:05:00', 1);
+insert into test_cc values('a','b', '2022-04-18 02:15:00', 3);
+insert into test_cc values('a','b', '2022-04-18 03:15:00', 1);
+insert into test_cc values('c','a', '2022-04-18 03:45:00', 1);
+insert into test_cc values('c','a', '2022-04-18 03:25:00', 2);
+insert into test_cc values('c','a', '2022-04-18 03:27:00', 3);
+
+select v2, count(1), count(distinct v1) from test_cc group by v2;
```
Comment (stdpain) [inline test/sql/test_agg_function/T/test_count_distinct:33]: add session variables `set new_planner_agg_stage=3` ?

Comment (Seaven) [reply test/sql/test_agg_function/T/test_count_distinct:33]: don't need

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java|8|8|100.00%|[]|

**Review by stdpain — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**